### PR TITLE
Utility and server: Get mod search paths from env MOD_SEARCH_PATHS

### PR DESCRIPTION
**Pull request by phrohdoh**
This will allow thirdparty game developers to write and invoke custom *and shipped* utility commands.

The mod template will want this so it can wrap the utility in a shell script like the example below.

```sh
# utility.sh in the mod template
MODLAUNCHER=$(python -c "import os; print(os.path.realpath('$0'))")
MOD_ROOT=$(dirname "$MODLAUNCHER")
ENGINE_ROOT="$MOD_ROOT/engine"

MOD_SEARCH_PATHS="$ENGINE_ROOT/mods,$MOD_ROOT/mods" mono ./engine/OpenRA.Utility.exe "$@"
```

```sh
$ ./utility.sh
Run `OpenRA.Utility.exe [MOD]` to see a list of available commands.
The available mods are: all, cnc, d2k, modchooser, ra, ts, ra2

$ ./utility.sh ra2 | grep "import.*map"
  --import-ra-map FILENAME
  --import-td-map FILENAME
  --import-ts-map FILENAME
```

```sh
thill@tmba /Users/.../openra $ MOD_SEARCH_PATHS=. orautil
Run `OpenRA.Utility.exe [MOD]` to see a list of available commands.
The available mods are:

thill@tmba /Users/.../openra $ MOD_SEARCH_PATHS=./mods orautil
Run `OpenRA.Utility.exe [MOD]` to see a list of available commands.
The available mods are: all, cnc, d2k, modchooser, ra, ts

thill@tmba /Users/.../openra $ orautil
Run `OpenRA.Utility.exe [MOD]` to see a list of available commands.
The available mods are: all, cnc, d2k, modchooser, ra, ts

thill@tmba /Users/.../openra-ra2 $ MOD_SEARCH_PATHS=$(pwd)/mods,$(pwd)/engine/mods orautil
Run `OpenRA.Utility.exe [MOD]` to see a list of available commands.
The available mods are: ra2, all, cnc, d2k, modchooser, ra, ts

```

### CODE REVIEW

File: OpenRA.Utility/Program.cs
Comment (pchote) [inline OpenRA.Utility/Program.cs]: The Engine.* args aren't stored in the settings, so this shouldn't need to (and IMO doesn't want to) change.

File: OpenRA.Utility/Program.cs
Comment (pchote) [inline OpenRA.Utility/Program.cs]: This skipped value approach doesn't sit well with me, it feels like something that tries/wants to be generic but is only ever going to be useful in this specific case.

IMO it would be a lot cleaner to search for `args.FirstOrDefault(a => a.StartsWith("Engine.ModSearchPaths="))` instead of shoehorning `Arguments` in here.

Comment (pchote) [inline OpenRA.Utility/Program.cs]: Another option would be to read `Engine.ModSearchPaths` from an environment variable, but we will need to think a bit about how we want to handle the other launcher -> game/utility/server bits first if we want to do this.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Utility/Program.cs
Comment (Phrohdoh) [inline OpenRA.Utility/Program.cs]: What does this section have to do with settings?
This is mapping cli args to Arguments.

File: OpenRA.Utility/Program.cs
Comment (Phrohdoh) [inline OpenRA.Utility/Program.cs]: If we go the `FirstOrDefault` route then the entire Arguments regex/dict is unnecessary.

File: OpenRA.Utility/Program.cs
Comment (pchote) [inline OpenRA.Utility/Program.cs]: `Game.InitializeSettings` populates the `Game.Settings` object with the contents of settings.yaml and the overrides from arguments.

**Comment by Phrohdoh** [comment]
This has been updated to get the search path from the environment variable `MOD_SEARCH_PATHS`.

File: OpenRA.Utility/Program.cs
Comment (pchote) [inline OpenRA.Utility/Program.cs]: It would be cleaner to use
```csharp
envModSearchPaths.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries)
	.Select(x => x.Trim())
	.ToArray()
```

and then drop the explicit loop.

**Comment by pchote** [comment]
Could you please make the same changes to the dedicated server exe?

**Comment by Phrohdoh** [comment]
@pchote done.

File: OpenRA.Server/Program.cs
Comment (pchote) [inline OpenRA.Server/Program.cs]: You can simplify this further by using FieldLoader:
```csharp
var envModSearchPaths = Environment.GetEnvironmentVariable("MOD_SEARCH_PATHS");
var modSearchPaths = !string.IsNullOrWhiteSpace(envModSearchPaths) ?
	FieldLoader.GetValue<string[]>("MOD_SEARCH_PATHS", envModSearchPaths) :
	new[] { Path.Combine(".", "mods"), Path.Combine("^", "mods") };
```

**Review by pchote — APPROVED** [review]
Works as advertised, and good to merge with the simplification suggested above

**Comment by pchote** [comment]
Can we please get a second review here?  This is holding up work on the mod template.

**Review by obrakmann — APPROVED** [review]
👍

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7566e7ebd505de5472bfb5ea726273e57d843741)

**Pull request merged**
